In [ ]:
import pandas as pd
train_df = pd.read_csv("../data/raw/UNSW_NB15_testing-set.csv")   # 175,341 rows
test_df = pd.read_csv("../data/raw/UNSW_NB15_training-set.csv")   # 82,332 rows

print("Train shape:", train_df.shape)
print("Test shape:", test_df.shape)

Train shape: (175341, 45)
Test shape: (82332, 45)


In [25]:
print("Train shape:", train_df.shape)
print("Test shape:", test_df.shape)

train_df.head()
train_df.columns.tolist()
train_df.info()

Train shape: (175341, 45)
Test shape: (82332, 45)
<class 'pandas.DataFrame'>
RangeIndex: 175341 entries, 0 to 175340
Data columns (total 45 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   id                 175341 non-null  int64  
 1   dur                175341 non-null  float64
 2   proto              175341 non-null  str    
 3   service            175341 non-null  str    
 4   state              175341 non-null  str    
 5   spkts              175341 non-null  int64  
 6   dpkts              175341 non-null  int64  
 7   sbytes             175341 non-null  int64  
 8   dbytes             175341 non-null  int64  
 9   rate               175341 non-null  float64
 10  sttl               175341 non-null  int64  
 11  dttl               175341 non-null  int64  
 12  sload              175341 non-null  float64
 13  dload              175341 non-null  float64
 14  sloss              175341 non-null  int64  
 15  dloss       

In [26]:
train_df.isnull().sum().sort_values(ascending=False)
train_df.duplicated().sum()
train_df.describe().T

,count,mean,std,min,25%,50%,75%,max
id,175341.0,8.767100e+04,5.061673e+04,1.0,43836.000000,87671.000000,1.315060e+05,1.753410e+05
dur,175341.0,1.359389e+00,6.480249e+00,0.0,0.000008,0.001582,6.680690e-01,5.999999e+01
spkts,175341.0,2.029866e+01,1.368876e+02,1.0,2.000000,2.000000,1.200000e+01,9.616000e+03
dpkts,175341.0,1.896959e+01,1.102583e+02,0.0,0.000000,2.000000,1.000000e+01,1.097400e+04
sbytes,175341.0,8.844844e+03,1.747656e+05,28.0,114.000000,430.000000,1.418000e+03,1.296523e+07
dbytes,175341.0,1.492892e+04,1.436542e+05,0.0,0.000000,164.000000,1.102000e+03,1.465555e+07
rate,175341.0,9.540619e+04,1.654010e+05,0.0,32.786140,3225.806520,1.250000e+05,1.000000e+06
sttl,175341.0,1.795470e+02,1.029400e+02,0.0,62.000000,254.000000,2.540000e+02,2.550000e+02
dttl,175341.0,7.960957e+01,1.105069e+02,0.0,0.000000,29.000000,2.520000e+02,2.540000e+02
sload,175341.0,7.345403e+07,1.883574e+08,0.0,13053.338870,879674.750000,8.888889e+07,5.988000e+09


In [27]:
train_df["label"].value_counts()
train_df["label"].value_counts(normalize=True) * 100

label
1    68.062233
0    31.937767
Name: proportion, dtype: float64

In [28]:
train_df["attack_cat"].value_counts()
train_df["attack_cat"].value_counts(normalize=True) * 100

attack_cat
Normal            31.937767
Generic           22.812691
Exploits          19.044605
Fuzzers           10.370649
DoS                6.994371
Reconnaissance     5.983198
Analysis           1.140635
Backdoor           0.995774
Shellcode          0.646169
Worms              0.074141
Name: proportion, dtype: float64

In [29]:
categorical_columns = ["proto", "service", "state"]

for column in categorical_columns:
    print(f"\n{column}")
    print(train_df[column].value_counts())


proto
proto
tcp       79946
udp       63283
unas      12084
arp        2859
ospf       2595
          ...  
rdp          98
netblt       98
igmp         18
icmp         15
rtp           1
Name: count, Length: 133, dtype: int64

service
service
-           94168
dns         47294
http        18724
smtp         5058
ftp-data     3995
ftp          3428
ssh          1302
pop3         1105
dhcp           94
snmp           80
ssl            56
irc            25
radius         12
Name: count, dtype: int64

state
state
INT    82275
FIN    77825
CON    13152
REQ     1991
RST       83
ECO       12
PAR        1
URN        1
no         1
Name: count, dtype: int64


In [30]:
import numpy as np
train_df.select_dtypes(include=np.number).describe().T

,count,mean,std,min,25%,50%,75%,max
id,175341.0,8.767100e+04,5.061673e+04,1.0,43836.000000,87671.000000,1.315060e+05,1.753410e+05
dur,175341.0,1.359389e+00,6.480249e+00,0.0,0.000008,0.001582,6.680690e-01,5.999999e+01
spkts,175341.0,2.029866e+01,1.368876e+02,1.0,2.000000,2.000000,1.200000e+01,9.616000e+03
dpkts,175341.0,1.896959e+01,1.102583e+02,0.0,0.000000,2.000000,1.000000e+01,1.097400e+04
sbytes,175341.0,8.844844e+03,1.747656e+05,28.0,114.000000,430.000000,1.418000e+03,1.296523e+07
dbytes,175341.0,1.492892e+04,1.436542e+05,0.0,0.000000,164.000000,1.102000e+03,1.465555e+07
rate,175341.0,9.540619e+04,1.654010e+05,0.0,32.786140,3225.806520,1.250000e+05,1.000000e+06
sttl,175341.0,1.795470e+02,1.029400e+02,0.0,62.000000,254.000000,2.540000e+02,2.550000e+02
dttl,175341.0,7.960957e+01,1.105069e+02,0.0,0.000000,29.000000,2.520000e+02,2.540000e+02
sload,175341.0,7.345403e+07,1.883574e+08,0.0,13053.338870,879674.750000,8.888889e+07,5.988000e+09


In [31]:
important_features = [
    "dur",
    "sbytes",
    "dbytes",
    "spkts",
    "dpkts",
    "rate"
]

train_df.groupby("label")[important_features].mean().T

label,0,1
dur,1.017177,1.519969
sbytes,4105.702929,11068.655357
dbytes,31049.463143,7364.456256
spkts,30.725482,15.405946
dpkts,38.057696,10.012619
rate,13799.312118,133699.690589


Exploratory Data Analysis Findings

1. Dataset Overview

The UNSW-NB15 dataset contains summarized network-flow records classified as normal or malicious traffic.

Dataset	Rows	Columns
Training set	175,341	45
Testing set	82,332	45

Each dataset contains:

39 numerical input features
3 categorical input features: proto, service, and state
1 identifier column: id
1 binary target: label
1 attack-category target: attack_cat

After removing id, label, and attack_cat, 42 input features remain for model training. The testing dataset will remain untouched until final evaluation.

2. Data Quality

Missing values: 0
Infinite numerical values: 0
Every training column contains 175,341 non-null values.
The - value in service represents an unidentified service and is treated as a valid category.
Repeated flow records will be retained for the initial baseline because identical summarized network flows may be legitimate observations.
Extreme numerical values will not be removed automatically because unusual traffic can represent genuine attacks.

3. Binary Target Distribution

Label	Meaning	Percentage
0	Normal traffic	31.94%
1	Attack traffic	68.06%

Attack traffic represents approximately two-thirds of the training dataset. Therefore, the binary target is moderately imbalanced.

The initial baseline will be trained without resampling. It will be evaluated using accuracy, precision, recall, F1-score, ROC-AUC, and a confusion matrix.

Attack recall is especially important because a false negative represents malicious traffic incorrectly classified as normal.

4. Attack-Category Distribution

Category	Percentage
Normal	31.9378%
Generic	22.8127%
Exploits	19.0446%
Fuzzers	10.3706%
DoS	6.9944%
Reconnaissance	5.9832%
Analysis	1.1406%
Backdoor	0.9958%
Shellcode	0.6462%
Worms	0.0741%

The attack-category target is highly imbalanced. Generic, Exploits, and Fuzzers are relatively common, while Analysis, Backdoor, Shellcode, and especially Worms are rare.

Accuracy alone will not reliably evaluate the attack-category classifier. Macro F1-score, weighted F1-score, per-class recall, and a confusion matrix should also be used.

5. Categorical Features

Protocol (proto)
Contains 133 unique categories.
tcp and udp are the most frequent protocols.
unas, arp, and ospf are also relatively common.
Several protocols occur very rarely.
The large number of protocol categories will increase the number of features created by one-hot encoding.
Service (service)
Contains 13 unique categories.
- is the most frequent value.
dns and http are the next most frequent services.
radius, irc, ssl, snmp, and dhcp are rare services.
The - value will be retained as a valid category.
Connection State (state)
Contains 9 unique categories.
INT, FIN, and CON are the most common connection states.
PAR, URN, and no occur only once.
Rare connection states will initially be retained.

The categorical features will be one-hot encoded. Unknown categories must be handled safely when processing validation, testing, or future real-world data.

6. Numerical Features

The numerical features describe different properties of network flows, including:

Flow duration
Source and destination packet counts
Source and destination byte counts
Traffic rate
Time-to-live values
Packet load and packet loss
Packet timing and jitter
TCP window and sequence information
HTTP and FTP activity
Connection-frequency statistics

These numerical features have significantly different value ranges. Therefore, they will be standardized when training Logistic Regression. Tree-based models generally do not require feature scaling.

7. Normal and Attack Traffic Comparison
Feature	Normal mean	Attack mean	Observation
dur	1.0172	1.5200	Attacks have a longer average duration
sbytes	4,105.70	11,068.66	Attacks have approximately 2.7 times more source bytes
dbytes	31,049.46	7,364.46	Normal flows have considerably more destination bytes
spkts	30.73	15.41	Normal flows contain more source packets
dpkts	38.06	10.01	Normal flows contain more destination packets
rate	13,799.31	133,699.69	Attacks have approximately 9.7 times the average traffic rate

These differences suggest that the features may help distinguish normal and malicious traffic. However, differences between group averages do not prove that a feature is predictive.

Extreme values can strongly affect averages, so feature distributions and median values should also be considered.

8. Leakage and Unusable Columns
Binary Intrusion-Detection Model

The binary model predicts whether a network flow is normal or malicious.

Target: label
Remove id because it is only a row identifier.
Remove label because it is the prediction target.
Remove attack_cat because it directly reveals whether the traffic is normal or malicious.
Attack-Category Model

The second model predicts the specific type of network attack.

Target: attack_cat
Use only records where label is equal to 1.
Remove id, label, and attack_cat from the model inputs.
Removing label prevents target information from leaking into the classifier.
9. Preprocessing Decisions

The preprocessing pipeline will:

Remove identifiers, target columns, and leakage columns.
Retain repeated network-flow records for the initial baseline.
Retain unusual numerical values unless they are proven invalid.
Retain - as a valid service category.
One-hot encode proto, service, and state.
Scale numerical features for Logistic Regression.
Safely handle previously unseen categorical values.
Fit preprocessing transformations using training data only.
Use a stratified training-validation split.
Keep the official testing dataset untouched until final evaluation.
10. Conclusion

The UNSW-NB15 dataset is suitable for developing the NetGuard-ML intrusion-detection system.

The binary target is moderately imbalanced, with attacks representing 68.06% of the training records. The attack-category target is highly imbalanced, with several rare attack types.

Categorical features require encoding, while numerical features require scaling for Logistic Regression. The id, label, and attack_cat columns must be handled carefully to prevent data leakage.

The next step is to create a reusable preprocessing pipeline and train the binary Logistic Regression baseline.

In [35]:
import numpy as np

duplicate_count = (
    train_df.drop(columns=["id"])
    .duplicated()
    .sum()
)

numeric_df = train_df.select_dtypes(include=np.number)
infinite_count = np.isinf(numeric_df).sum().sum()

print("Repeated rows without ID:", duplicate_count)
print("Infinite numerical values:", infinite_count)

Repeated rows without ID: 67601
Infinite numerical values: 0


In [32]:
import numpy as np

numeric_df = train_df.select_dtypes(include=np.number)

infinite_count = np.isinf(numeric_df).sum().sum()
print("Total infinite values:", infinite_count)

Total infinite values: 0


In [33]:
duplicate_count = train_df.drop(columns=["id"]).duplicated().sum()
print("Duplicate rows:", duplicate_count)

Duplicate rows: 67601


In [34]:
print(train_df.isnull().sum().sum())

0
